File: src/main/java/net/arvandor/numinoustreasury/command/node/session/NodeSessionDropTableCommand.java
package net.arvandor.numinoustreasury.command.node.session;

import static net.md_5.bungee.api.ChatColor.*;
import static net.md_5.bungee.api.chat.ClickEvent.Action.RUN_COMMAND;
import static net.md_5.bungee.api.chat.HoverEvent.Action.SHOW_TEXT;

import com.rpkit.core.bukkit.pagination.PaginatedView;
import com.rpkit.core.service.Services;
import net.arvandor.numinoustreasury.droptable.NuminousDropTable;
import net.arvandor.numinoustreasury.droptable.NuminousDropTableService;
import net.arvandor.numinoustreasury.node.NuminousNodeCreationSession;
import net.arvandor.numinoustreasury.node.NuminousNodeService;
import net.md_5.bungee.api.chat.*;
import net.md_5.bungee.api.chat.hover.content.Text;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public final class NodeSessionDropTableCommand implements CommandExecutor, TabCompleter {

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!sender.hasPermission("numinoustreasury.command.node.create")) {
            sender.sendMessage(RED + "You do not have permission to create nodes.");
            return true;
        }
        if (!(sender instanceof Player player)) {
            sender.sendMessage(RED + "You must be a player to create nodes.");
            return true;
        }
        NuminousNodeService nodeService = Services.INSTANCE.get(NuminousNodeService.class);
        NuminousNodeCreationSession session = nodeService.getNodeCreationSession(player);
        if (session == null) {
            session = new NuminousNodeCreationSession();
            nodeService.setNodeCreationSession(player, session);
        }
        NuminousDropTableService dropTableService = Services.INSTANCE.get(NuminousDropTableService.class);
        boolean dropTableSpecified = args.length > 1 && args[0].equalsIgnoreCase("set");
        int page = 1;
        if (args.length > 1 && args[0].equalsIgnoreCase("page")) {
            try {
                page = Integer.parseInt(args[1]);
            } catch (NumberFormatException exception) {
                sender.sendMessage(RED + "Page number must be an integer.");
                return true;
            }
        }
        if (dropTableSpecified) {
            String dropTableId = Arrays.stream(args).skip(1).collect(Collectors.joining(" "));
            NuminousDropTable dropTable = dropTableService.getDropTableById(dropTableId);
            if (dropTable == null) {
                sender.sendMessage(RED + "There is no drop table with that ID.");
                return true;
            }
            session.setDropTable(dropTable);
            session.display(player);
        } else {
            List<NuminousDropTable> dropTables = dropTableService.getDropTables();
            if (dropTables.isEmpty()) {
                sender.sendMessage(RED + "There are no drop tables configured.");
                return true;
            }
            TextComponent title = new TextComponent("Select a drop table: ");
            title.setColor(WHITE);
            PaginatedView view = PaginatedView.fromChatComponents(
                    new BaseComponent[] {
                            title
                    },
                    dropTables.stream().map(dropTable -> new ComponentBuilder(dropTable.getId())
                            .color(GREEN)
                            .event(new ClickEvent(
                                    RUN_COMMAND,
                                    "/node session droptable set " + dropTable.getId()
                            ))
                            .event(new HoverEvent(
                                    SHOW_TEXT,
                                    new Text(
                                            new ComponentBuilder("Click here to set the drop table to " + dropTable.getId())
                                                    .color(GRAY)
                                                    .create()
                                    )
                            ))
                            .create()).toList(),
                    GREEN + "< Previous",
                    "Click here to view the previous page",
                    GREEN + "Next >",
                    "Click here to view the next page",
                    (pageNumber) -> "Page " + pageNumber,
                    10,
                    (pageNumber) -> "/node session droptable page " + pageNumber
            );
            if (view.isPageValid(page)) {
                view.sendPage(sender, page);
            } else {
                sender.sendMessage(RED + "Invalid page number.");
            }
        }
        return true;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        NuminousDropTableService dropTableService = Services.INSTANCE.get(NuminousDropTableService.class);
        if (args.length == 0) {
            return dropTableService.getDropTables().stream()
                    .map(NuminousDropTable::getId)
                    .toList();
        } else {
            return dropTableService.getDropTables().stream()
                    .map(NuminousDropTable::getId)
                    .filter(id -> id.toLowerCase().startsWith(String.join(" ", args).toLowerCase()))
                    .toList();
        }
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/node/session/NodeSessionDropTableCommand.kt
package net.arvandor.numinoustreasury.command.node.session

import com.rpkit.core.bukkit.pagination.PaginatedView
import com.rpkit.core.service.Services
import net.arvandor.numinoustreasury.droptable.NuminousDropTableService
import net.arvandor.numinoustreasury.node.NuminousNodeCreationSession
import net.arvandor.numinoustreasury.node.NuminousNodeService
import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.chat.BaseComponent
import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.ClickEvent.Action.RUN_COMMAND
import net.md_5.bungee.api.chat.ComponentBuilder
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.HoverEvent.Action.SHOW_TEXT
import net.md_5.bungee.api.chat.TextComponent
import net.md_5.bungee.api.chat.hover.content.Text
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

class NodeSessionDropTableCommand : CommandExecutor, TabCompleter {
    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (!sender.hasPermission("numinoustreasury.command.node.create")) {
            sender.sendMessage(ChatColor.RED.toString() + "You do not have permission to create nodes.")
            return true
        }
        if (sender !is Player) {
            sender.sendMessage(ChatColor.RED.toString() + "You must be a player to create nodes.")
            return true
        }
        val nodeService =
            Services.INSTANCE.get(
                NuminousNodeService::class.java,
            )
        var session = nodeService.getNodeCreationSession(sender)
        if (session == null) {
            session = NuminousNodeCreationSession()
            nodeService.setNodeCreationSession(sender, session)
        }
        val dropTableService =
            Services.INSTANCE.get(
                NuminousDropTableService::class.java,
            )
        val dropTableSpecified = args.size > 1 && args[0].equals("set", ignoreCase = true)
        var page = 1
        if (args.size > 1 && args[0].equals("page", ignoreCase = true)) {
            try {
                page = args[1].toInt()
            } catch (exception: NumberFormatException) {
                sender.sendMessage(ChatColor.RED.toString() + "Page number must be an integer.")
                return true
            }
        }
        if (dropTableSpecified) {
            val dropTableId = args.drop(1).joinToString(" ")
            val dropTable = dropTableService.getDropTableById(dropTableId)
            if (dropTable == null) {
                sender.sendMessage(ChatColor.RED.toString() + "There is no drop table with that ID.")
                return true
            }
            session.dropTable = dropTable
            session.display(sender)
        } else {
            val dropTables = dropTableService.dropTables
            if (dropTables.isEmpty()) {
                sender.sendMessage(ChatColor.RED.toString() + "There are no drop tables configured.")
                return true
            }
            val title = TextComponent("Select a drop table: ")
            title.color = ChatColor.WHITE
            val view =
                PaginatedView.fromChatComponents(
                    arrayOf<BaseComponent>(
                        title,
                    ),
                    dropTables.map { dropTable ->
                        ComponentBuilder(
                            dropTable.id,
                        )
                            .color(ChatColor.GREEN)
                            .event(
                                ClickEvent(
                                    RUN_COMMAND,
                                    "/node session droptable set " + dropTable.id,
                                ),
                            )
                            .event(
                                HoverEvent(
                                    SHOW_TEXT,
                                    Text(
                                        ComponentBuilder("Click here to set the drop table to " + dropTable.id)
                                            .color(ChatColor.GRAY)
                                            .create(),
                                    ),
                                ),
                            )
                            .create()
                    },
                    ChatColor.GREEN.toString() + "< Previous",
                    "Click here to view the previous page",
                    ChatColor.GREEN.toString() + "Next >",
                    "Click here to view the next page",
                    { pageNumber: Int -> "Page $pageNumber" },
                    10,
                    { pageNumber: Int -> "/node session droptable page $pageNumber" },
                )
            if (view.isPageValid(page)) {
                view.sendPage(sender, page)
            } else {
                sender.sendMessage(ChatColor.RED.toString() + "Invalid page number.")
            }
        }
        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String>? {
        val dropTableService =
            Services.INSTANCE.get(
                NuminousDropTableService::class.java,
            )
        return if (args.isEmpty()) {
            dropTableService.dropTables
                .map { obj -> obj.id }
        } else {
            dropTableService.dropTables
                .map { obj -> obj.id }
                .filter { id: String ->
                    id.lowercase()
                        .startsWith(args.joinToString(" ").lowercase())
                }
        }
    }
}
